### ISSUE TITLE
Disk quotas are showing action button for bind mounted folders

### ISSUE THREAD

**Issue by adbario** [webmin/webmin#706]
Hi Jamie,

there was a previous issue about this (#493), which was marked as solved, but this still seems to be happening using latest stable Webmin with CentOS 7.4

<img width="1031" alt="screen shot 2017-11-27 at 21 38 35" src="https://user-images.githubusercontent.com/22136575/33290712-7e9acc24-d3bb-11e7-843f-6e4426ee26f0.png">

/data is an actual drive which has vhosts and mysql folders that are mounted with bind to /var/www/vhosts and /var/lib/mysql. These bind-mounted folders shouldn't have action button to enable quotas.

Also very related to this issue, is that Virtualmin doesn't recognize quotas on these bind mounts correctly as stated here https://www.virtualmin.com/node/34301. There's a suggested fix by you which works very well, but it gets overwritted when running "Re-check Configuration".

**Comment by jcameron** [comment]
So in this case, since those are bind mounts I assume the correct behavior is for those filesystems to not show up in the Disk Quotas module at all?

**Comment by adbario** [comment]
Either not show up at all or at least hide the "Enable quotas" button as it's not doing anything for these mounts. Apparently the previous fix worked for some other OS but doesn't seem to cover CentOS.

Also tried to look at the sources for this Virtualmin issue, but my coding skills sadly doesnät cover Perl so couldn't find out a way to try to fix it. Thought about creating another issue for this, but these are probably related.

**Comment by qooob** [comment]
@adbario What is the output of:

```
findmnt 
```
and 

```
findmnt | grep -oP \'\[\K[^\]]+\'
```

**Comment by adbario** [comment]
@iliaross Here's the output:

```
$ findmnt
/                                     /dev/vda3  xfs        rw,relatime,seclabel,attr2,inode64,noquota
...
├─/data                               /dev/vdb1  ext4       rw,relatime,seclabel,quota,usrquota,grpquota,data=ordered
├─/var/www/vhosts                     /dev/vdb1[/vhosts]
                                                 ext4       rw,relatime,seclabel,quota,usrquota,grpquota,data=ordered
└─/var/lib/mysql                      /dev/vdb1[/mysql]
                                                 ext4       rw,relatime,seclabel,quota,usrquota,grpquota,data=ordered
```

```
$ findmnt | grep -oP \'\[\K[^\]]+\'
```
Outputs nothing for me.

**Comment by qooob** [comment]
When I get home I'll take a look.

Stay tuned.

**Comment by adbario** [comment]
Thanks, appreciate it!

**Comment by qooob** [comment]
@adbario Sorry, I mean to ask what is the output of?

```
findmnt | grep -oP '\[\K[^\]]+'
```

**Comment by adbario** [comment]
That outputs the mounted folders:
```
$ findmnt | grep -oP '\[\K[^\]]+'
/vhosts
/mysql
```

**Comment by qooob** [comment]
Are these all that have to be ignored?

**Comment by adbario** [comment]
These are the only two bind mounted folders, so yes. You can't turn on quotas specifically on these anyway, only for the main drive which is shown correctly already.

**Comment by qooob** [comment]
I know. That was exactly what I was fixing for Fedora. It should just work. And output is correct.

What are your Webmin and Virtualmin versions?

**Comment by qooob** [comment]
I just checked and in case I remove my commit, I again see bind mounts in the list.

Make sure that you have it in your `/usr/libexec/webmin/quota/linux-lib.pl`:

```Perl
sub quota_can
{
my %exclude_mounts;
if (&has_command("findmnt")) {
	%exclude_mounts = map { $_ => 1 } split( /\n/m, backquote_command('findmnt | grep -oP \'\[\K[^\]]+\'') );
	}
    
# Not possible on bind mounts
if ($_[0]->[2] =~ /^bind/ || exists($exclude_mounts{$_[0]->[0]})) {
	return 0;
	}

return ( $_[1]->[3] =~ /usrquota|usrjquota/ || $_[0]->[3] =~ /usrquota|usrjquota/ ? 1 : 0 ) +
       ( $_[1]->[3] =~ /grpquota|grpjquota/      || $_[0]->[3] =~ /grpquota|grpjquota/ ? 2 : 0 );
}

```

Lines 83-97.

**Comment by adbario** [comment]
Also they show up only after enabling quotas for the disk "/data" itself.

**Comment by qooob** [comment]
Probably Jamie knows the theory. Let's wait and see. Other than that, I personally could only fix it in practice for now, by trying to reproduce it.

**Comment by adbario** [comment]
Thanks for your efforts @iliaross

So the "exists($exclude_mounts{$_[0]->[0]})" function doesn't work properly I guess. Printing the contents of "%exclude_mounts" gives a long list, all separated by "1":
```
/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1/mysql1/vhosts1
```
"$_[0]->[0]" contains:
```
/sys/proc/dev/sys/kernel/security/dev/shm/dev/pts/run/sys/fs/cgroup/sys/fs/cgroup/systemd/sys/fs/pstore/sys/fs/cgroup/net_cls,net_prio/sys/fs/cgroup/hugetlb/sys/fs/cgroup/cpuset/sys/fs/cgroup/pids/sys/fs/cgroup/devices/sys/fs/cgroup/cpu,cpuacct/sys/fs/cgroup/blkio/sys/fs/cgroup/memory/sys/fs/cgroup/freezer/sys/fs/cgroup/perf_event/sys/kernel/config//sys/fs/selinux/sys/kernel/debug/dev/mqueue/dev/hugepages/data/var/www/vhosts/var/lib/mysql/boot/run/user/0swap
```

Is the Virtualmin issue happening for same reasons?

**Comment by qooob** [comment]
I see, I think `=>` fat arrow operator is not supported such logic in _Perl_ prior to 5.20 and your OS has 5.16.. Brrr.

We need to convert the code. It will take some time for me, I'm not Perl expert. I will call for @swelljoe for the rescue, unless Jamie posts a fix sooner.

**Comment by swelljoe** [comment]
=> and -> have been in perl literally forever. It's something else. I'm reading over the issue now.

**Comment by adbario** [comment]
When manually setting the full path to one of the directories, it won't show up:
```
%exclude_mounts = "/var/lib/mysql";
```
So I guess exclude_mounts should have the full target paths, while now it's having the original paths.

**Comment by qooob** [comment]
To make this clearer here is what output %exclude_mounts on my Perl 5.24:

```
'/var/spool/mail' => 1, '/srv' => 1, '/var/lib/mysql' => 1, '/var/www' => 1
```

**Comment by qooob** [comment]
The same outputs:

```
findmnt | grep -oP '\[\K[^\]]+'
```

**Comment by adbario** [comment]
Maybe that's the issue, for you the output contains full paths, for me it has only the lowest folder or paths in original location:
```
/vhosts
/mysql
```
while they should be:
```
/var/www/vhosts
/var/lib/mysql
```

**Comment by qooob** [comment]
That what I was asking you above. Once it's not the full path, regex has to be fixed.

When you run in console:
```
findmnt | grep -oP '\[\K[^\]]+'
```

You need to make sure that it's returned as:
```Bash
[root@server ~]# findmnt | grep -oP '\[\K[^\]]+'
/var/spool/mail
/var/www
/var/lib/mysql
/srv
```

**Comment by qooob** [comment]
Looking at the plain output of `findmnt` I can see the difference now:

It returns on CentOS 
```
/dev/vdb1[/vhosts]
```

Instead of what would be on Fedora
```
/dev/vdb1[/var/www/vhosts]
```

So it's not the regex. I'll check if we can force `findmnt` to display full path for bind mounts..

**Comment by adbario** [comment]
Sorry about it, I didn't get that you meant the final folders. Well at least we know where the issue is now.

**Comment by qooob** [comment]
@adbario Could you please try running `findmnt` with some keys from below:

```
-F, --tab-file <path>  alternative file for --fstab, --mtab or --kernel options
 -f, --first-only       print the first found filesystem only
 -i, --invert           invert the sense of matching
 -l, --list             use list format output
 -N, --task <tid>       use alternative namespace (/proc/<tid>/mountinfo file)
 -n, --noheadings       don't print column headings
 -u, --notruncate       don't truncate text in columns
 -O, --options <list>   limit the set of filesystems by mount options
 -o, --output <list>    the output columns to be shown
 -P, --pairs            use key="value" output format
 -r, --raw              use raw output format
 -t, --types <list>     limit the set of filesystems by FS types
 -v, --nofsroot         don't print [/dir] for bind or btrfs mounts
 -R, --submounts        print all submounts for the matching filesystems

```

Make sure that it print the full path on bind mount.

**Comment by adbario** [comment]
The most useful ones could be maybe --list and --pairs. List would be without the tree and pairs will output:
```
...
TARGET="/data" SOURCE="/dev/vdb1" FSTYPE="ext4" OPTIONS="rw,relatime,seclabel,quota,usrquota,grpquota,data=ordered"
TARGET="/var/www/vhosts" SOURCE="/dev/vdb1[/vhosts]" FSTYPE="ext4" OPTIONS="rw,relatime,seclabel,quota,usrquota,grpquota,data=ordered"
TARGET="/var/lib/mysql" SOURCE="/dev/vdb1[/mysql]" FSTYPE="ext4" OPTIONS="rw,relatime,seclabel,quota,usrquota,grpquota,data=ordered"
TARGET="/boot" SOURCE="/dev/vda1" FSTYPE="xfs" OPTIONS="rw,relatime,seclabel,attr2,inode64,noquota"
...
```

**Comment by swelljoe** [comment]
I think -r/--raw looks easiest to parse. But, any of those options is better than the awful tree.

**Comment by qooob** [comment]
I would change regex:

```
\[\K[^\]]+
```

From the one above to the one that would get the first column with actual path.

**Comment by qooob** [comment]
@adbario It's very strange that you still don't have full path in the brackets.

The strange thing is, that I do on my CentOS!? How do you bind mount your devices?

**Comment by adbario** [comment]
This is how they are mounted:
```
$ cat /etc/fstab
...
/dev/vdb1 /data ext4 usrquota,grpquota 0 0
/data/vhosts /var/www/vhosts none bind 0 0
/data/mysql /var/lib/mysql none bind 0 0
```

**Comment by qooob** [comment]
I see now.

I think that when they mounted using system call via fstab, they are displayed differently.

Try to unmount it and then mount it using console:

```
mount --bind /data/mysql /var/lib/mysql
```

What's the output of `findmnt`?

**Comment by adbario** [comment]
Still the output is same:
```
$ umount /var/www/vhosts
$ mount --bind /data/vhosts /var/www/vhosts
$ findmnt -r
...
/var/www/vhosts /dev/vdb1[/vhosts] ext4 rw,relatime,seclabel,quota,usrquota,grpquota,data=ordered
...
```

**Comment by qooob** [comment]
😮 

What?

My CentOS 7.4 `findmnt` version is:

```
findmnt --version
```

What is yours?

**Comment by adbario** [comment]
```
findmnt from util-linux 2.23.2
```
Anyway, these folders shold definitely be mounted on boot so the issue still would exist.

**Comment by qooob** [comment]
Okay, then we'll have to find other way to grab first column. It's not that difficult. 

I will take a look in case neither Joe nor Jamie would.

Thank you for helping making Webmin better! 😉

**Comment by adbario** [comment]
This regex works for capturing the full path, catches the first string from the beginning of line before space when line contains "[/":

https://regex101.com/r/fyfjnU/3
```
$ findmnt -r | grep -oP '^(\S+)(?=.*\[\/)'

/var/lib/mysql
/var/www/vhosts
```

So changing linux-lib.pl like this works perfectly for me:
```perl
if (&has_command("findmnt")) {
  %exclude_mounts = map { $_ => 1 } split( /\n/m, backquote_command('findmnt -r | grep -oP \'^(\S+)(?=.*\[\/)\'') );
}
```
Could you guys try this out?

**Comment by qooob** [comment]
Your example works fine. I expect lookaheads are supported in Perl since longs time as well 😉 .

I made Webmin patch:
https://github.com/webmin/webmin/pull/708

As well as related Virtualmin patch:
[0900fdc](https://github.com/virtualmin/virtualmin-gpl/commit/0900fdc083e563754e56afc69e9879a2dad76202#diff-d93c5e3420554c840d321dbc5e2fca0b)

It will be fixed in upcoming Webmin 1.870.

**Comment by adbario** [comment]
Great!

Virtualmin fix doesn't cover the configuration check though:
Re-check configuration...
**Quotas are not enabled on the filesystem /var/www/vhosts which contains home directories under /var/www/vhosts and email files under /var/www/vhosts. Quota editing has been disabled.**

These troubles could easily be solved by partitioning disk and mounting the partitions where ever you need them, but bind mounting folders might give a bit more flexibility as you don't need to decide the sizes beforehand. Although with bind mounting the directory tree exists on two or more places in the directory hierarchy.
